kernel/src/main/scala/cats/kernel/instances/either.scala

```diff
@@ -63,21 +63,36 @@ trait EitherInstances0 extends EitherInstances1 {
           }
         }
     }
-}
 
-trait EitherInstances1 {
-  implicit def catsStdEqForEither[A, B](implicit A: Eq[A], B: Eq[B]): Eq[Either[A, B]] =
-    new Eq[Either[A, B]] {
-      def eqv(x: Either[A, B], y: Either[A, B]): Boolean =
+  implicit def catsStdHashForEither[A, B](implicit A: Hash[A], B: Hash[B]): Hash[Either[A, B]] =
+    new EitherEq[A, B](A, B) with Hash[Either[A, B]] {
+      def hash(x: Either[A, B]): Int = {
         x match {
-          case Left(xx) => y match {
-            case Left(yy) => A.eqv(xx, yy)
-            case Right(_) => false
-          }
-          case Right(xx) => y match {
-            case Left(_) => false
-            case Right(yy) => B.eqv(xx, yy)
-          }
+          case Left(xx) => Hash.caseClass1Hash("Left".##, A.hash(xx))
```

Comment:
could we cache `"Left".##` as a val and not recompute it each time?

Reply:
OK. (I thought that the compiler automatically does this because `"Left".##` is a constant, but sure, I'm going to write it as a val).

Reply:
Now no allocation and fully compliant with the `case class` hashCode.